src/libraries/Common/tests/System/Diagnostics/DebuggerAttributes.cs

```diff
@@ -86,18 +93,52 @@ public static IEnumerable<PropertyInfo> GetDebuggerVisibleProperties(Type debugg
 
         public static Type GetProxyType(Type type) => GetProxyType(type, type.GenericTypeArguments);
 
-        private static Type GetProxyType(Type type, Type[] genericTypeArguments)
+        internal static DebuggerDisplayResult ValidateFullyDebuggerDisplayReferences(object obj)
+        {
+            CustomAttributeData cad = FindAttribute(obj.GetType(), attributeType: typeof(DebuggerDisplayAttribute));
+
+            // Get the text of the DebuggerDisplayAttribute
+            string attrText = (string)cad.ConstructorArguments[0].Value;
+            string formattedValue = EvaluateDisplayString(attrText, obj);
+
+            string formattedKey = FormatDebuggerDisplayNamedArgument(nameof(DebuggerDisplayAttribute.Name), cad, obj);
+            string formattedType = FormatDebuggerDisplayNamedArgument(nameof(DebuggerDisplayAttribute.Type), cad, obj);
+
+            return new DebuggerDisplayResult { Value = formattedValue, Key = formattedKey, Type = formattedType };
+        }
+
+        internal static string ValidateDebuggerDisplayReferences(object obj)
+        {
+            CustomAttributeData cad = FindAttribute(obj.GetType(), attributeType: typeof(DebuggerDisplayAttribute));
+
+            // Get the text of the DebuggerDisplayAttribute
+            string attrText = (string)cad.ConstructorArguments[0].Value;
+
+            return EvaluateDisplayString(attrText, obj);
+        }
+
+        private static CustomAttributeData FindAttribute(Type type, Type attributeType)
         {
-            // Get the DebuggerTypeProxyAttribute for obj
-            CustomAttributeData[] attrs =
-                type.GetTypeInfo().CustomAttributes
-                .Where(a => a.AttributeType == typeof(DebuggerTypeProxyAttribute))
-                .ToArray();
-            if (attrs.Length != 1)
+            for (var t = type; t != null; t = t.BaseType)
             {
-                throw new InvalidOperationException($"Expected one DebuggerTypeProxyAttribute on {type}.");
+                CustomAttributeData[] attributes = t.GetTypeInfo().CustomAttributes
+                    .Where(a => a.AttributeType == attributeType)
+                    .ToArray();
+                if (attributes.Length != 0)
+                {
+                    if (attributes.Length > 1)
+                    {
+                        throw new InvalidOperationException($"Expected one {attributeType.Name} on {type} but found more.");
```

Comment:
Should this be an assert instead?

Reply:
IMO helper methods such as this one should not determine the outcome of a test (unless it's their main purpose). Test methods can handle exceptions (or not) based on their requirements. These tests worked this way previously, I haven't change it.